bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+			 prefix4_descr->ospf_route_type);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(prefix4_descr->flags,
+		       BGP_NLRI_TLV_PREFIX_DESCR_IP_REACHABILITY)) {
+		snprintfrr(buf, size, "%sIPv4:%pI4/%u", first ? "" : " ",
+			   &prefix4_descr->ip_reachability_prefix,
+			   prefix4_descr->ip_reachability_prefixlen);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(prefix4_descr->flags, BGP_NLRI_TLV_PREFIX_DESCR_MT_ID)) {
+		size = bgp_linkstate_nlri_mtid_display(&prefix4_descr->mtid,
+						       buf, size, first);
+		buf += strlen(buf);
+	}
+
+	if (!multiline) {
+		snprintf(buf, size, "}");
+		size -= strlen(buf);
+	}
+
+	return size;
+}
+
+static uint16_t bgp_linkstate_nlri_prefix4_prefixlen(
+	struct bgp_linkstate_type_prefix4 *p_prefix4)
+{
+	return sizeof(struct bgp_linkstate_type_prefix4) +
+	       p_prefix4->prefix_descr.mtid.length;
+}
+
+static int
+bgp_linkstate_nlri_prefix4_decode(struct bgp_linkstate_type_prefix4 **pp,
+				  uint8_t *pnt, uint16_t len)
+{
+	struct bgp_linkstate_type_prefix4 *p_prefix4 = *pp;
+	int ret = BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+	uint8_t *lim = pnt + len;
+	uint16_t type, length;
+	uint16_t *dst, *src;
+
+	p_prefix4->proto = pnt_decode8(&pnt);
+	p_prefix4->identifier = pnt_decode64(&pnt);
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+
+		if (!bgp_ls_tlv_check_size(type, length))
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+
+		switch (type) {
+		case BGP_LS_TLV_LOCAL_NODE_DESCRIPTORS:
+			if (p_prefix4->local_node_descr.flags != 0)
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_prefix4->local_node_descr.flags,
+				 BGP_NLRI_TLV_NODE_DESCR_LOCAL_NODE);
+			ret = bgp_linkstate_nlri_node_descriptor_decode(
+				&p_prefix4->local_node_descr, pnt, length);
+			break;
+		case BGP_LS_TLV_OSPF_ROUTE_TYPE:
+		case BGP_LS_TLV_IP_REACHABILITY_INFORMATION:
+			ret = bgp_linkstate_nlri_prefix4_descriptor_decode(
+				&p_prefix4->prefix_descr, pnt, type, length);
+			break;
+		case BGP_LS_TLV_MULTI_TOPOLOGY_ID:
+			if (CHECK_FLAG(p_prefix4->prefix_descr.flags,
+				       BGP_NLRI_TLV_PREFIX_DESCR_MT_ID))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			if ((length % 2) != 0)
+				/* Length must be a multiple of 2 */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_prefix4->prefix_descr.flags,
+				 BGP_NLRI_TLV_PREFIX_DESCR_MT_ID);
+			*pp = XREALLOC(
+				MTYPE_BGP_LS_TYPE_PREFIX4, *pp,
+				sizeof(struct bgp_linkstate_type_prefix4) +
+					length);
+			p_prefix4 = *pp;
+			p_prefix4->prefix_descr.mtid.length = length;
+			dst = (uint16_t *)&p_prefix4->prefix_descr.mtid.data;
+			src = (uint16_t *)pnt;
+			for (int i = 0; i < (length / 2); i++)
+				dst[i] = ntohs(src[i]);
+			break;
+		default:
+			zlog_err("%s: received an unexpected TLV type %hu",
```

Comment:
dont return error if unknown tlv.
- record the TLV type, and return NLRI_PARSE_OK; 
- use conditional debugging to output the message.

Reply:
see https://github.com/FRRouting/frr/pull/12649#discussion_r1169925019